src/stream/src/executor/watermark_filter.rs

```diff
@@ -0,0 +1,477 @@
+                            ctx.on_compute_error(err, &info.identity)
+                        });
+
+                    if let Some(output_chunk) = SimpleFilterExecutor::filter(chunk, pred_output)? {
+                        yield Message::Chunk(output_chunk);
+                    };
+
+                    yield Message::Watermark(Watermark::new(
+                        event_time_col_idx,
+                        current_watermark.clone(),
+                    ));
+                }
+                Message::Watermark(watermark) => {
+                    if watermark.col_idx != event_time_col_idx {
+                        yield Message::Watermark(watermark)
+                    } else {
+                        bail!("WatermarkFilterExecutor should not receive a watermark on the event it is filtering.")
+                    }
+                }
+                Message::Barrier(barrier) => {
+                    // Update the vnode bitmap for state tables of all agg calls if asked.
+                    if let Some(vnode_bitmap) = barrier.as_update_vnode_bitmap(ctx.id) {
+                        let previous_vnode_bitmap = table.update_vnode_bitmap(vnode_bitmap.clone());
+
+                        // Take the global max watermark when scaling happens.
+                        if previous_vnode_bitmap != vnode_bitmap {
+                            current_watermark =
+                                Self::get_global_max_watermark(&table, watermark_type.clone())
+                                    .await?;
+                        }
+                    }
+
+                    if barrier.checkpoint {
+                        // Persist the watermark when checkpoint arrives.
+                        let vnodes = table.get_vnodes();
+                        for vnode in vnodes.iter_pos() {
+                            let pk = Some(ScalarImpl::Int16(vnode as _));
+                            let row = Row::new(vec![pk, current_watermark.clone()]);
+                            table.insert(row);
+                        }
+                        table.commit(barrier.epoch).await?;
+                    } else {
+                        table.commit_no_data_expected(barrier.epoch);
+                    }
+
+                    yield Message::Barrier(barrier);
+                }
+            }
+        }
+    }
+
+    fn build_watermark_filter_expr(
+        watermark_type: DataType,
+        event_time_col_idx: usize,
+        watermark: Datum,
+    ) -> ExprResult<BoxedExpression> {
+        new_binary_expr(
+            Type::GreaterThanOrEqual,
+            DataType::Boolean,
+            InputRefExpression::new(watermark_type.clone(), event_time_col_idx).boxed(),
+            LiteralExpression::new(watermark_type, watermark).boxed(),
+        )
+    }
+
+    async fn get_global_max_watermark(
+        table: &StateTable<S>,
+        watermark_type: DataType,
+    ) -> StreamExecutorResult<Datum> {
```

Comment:
Should we use `ScalarImpl` here, or what's the semantic for watermark being `NULL`? cc @st1page.

Reply:
Agree, in the currently watermark filter design it should not be Null

Reply:
But maybe do it in later PRs.

Reply:
#6231